### PR TITLE
CORE-8891 encumbrance verification

### PR DESCRIPTION
**Pull request by blsemo**
Add the concept of an Encubrance group - encumbrance tag plus number of states in the group.

When verifying a transaction, add a check that either no encumbered states or the whole encumbrance group has been added as inputs.

API PR: https://github.com/corda/corda-api/pull/739

### CODE REVIEW

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/ContractStateAndEncumbranceTag.kt [hunk 1]
```diff
@@ -9,7 +10,10 @@ import net.corda.v5.ledger.utxo.TransactionState
 @CordaSerializable
 data class ContractStateAndEncumbranceTag(val contractState: ContractState, val encumbranceTag: String?) {
 
-    fun toTransactionState(notary: Party): TransactionState<*> {
-        return TransactionStateImpl(contractState, notary, encumbranceTag)
+    fun toTransactionState(notary: Party, encumbranceGroupSize: Int?): TransactionState<*> {
+        return TransactionStateImpl(contractState, notary, encumbranceTag?.let{
+            require(encumbranceGroupSize != null)
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/ContractStateAndEncumbranceTag.kt]: requireNotNull is a bit more concise

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    // group input by transaction id (encumbrance is only unique within one transaction output)
+    inputStateAndRefs.groupBy { it.ref.transactionHash }.forEach { statesByTx ->
+        // Filter out unencumbered states
+        statesByTx.value.filter { it.state.encumbrance != null }
+            // within each tx, group by encumbrance tag, store the output index and the encumbrance group size
+            .groupBy({ it.state.encumbrance!!.tag }, { Pair(it.ref.index, it.state.encumbrance!!.size) })
+            // for each output entry, run the checks
+            .forEach { encumbranceGroup ->
+                // Check that no input states have been duplicated to fool our counting
+                encumbranceGroup.value.groupBy { it.first }.filter { it.value.size > 1 }.apply{
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt]: Can we use parameters with names instead of `it` as its a bit hard to follow.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                    this.forEach{ failureReasons.add(
+                        ContractVerificationFailureImpl(
+                            contractClassName = "",
+                            contractStateClassNames = emptyList(),
+                            exceptionClassName = IllegalArgumentException::class.java.canonicalName,
+                            exceptionMessage = "Encumbrance check failed: State ${statesByTx.key}, " +
+                                    "${it.key} is part " +
+                                    "is used ${it.value.size} times as input!"))
+                    }
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt]: Looks like it could do with reformatting.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                }
+
+                val numberOfStatesPresent = encumbranceGroup.value.size
+                // if the size of the encumbrance group does not match the number of input states,
+                // then add a failure reason.
+                encumbranceGroup.value.forEach { encumbrancePair ->
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt]: Can do `encumbranceGroup.value.forEach { outputIndex, encumbranceGroupSize) ->` I think those are the right variables at least.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                                        "${encumbrancePair.first} is part " +
+                                        "of encumbrance group ${encumbranceGroup.key}, but only " +
+                                        "$numberOfStatesPresent states out of " +
+                                        "${encumbrancePair.second} encumbered states are present as inputs."
+
+                            )
+                        )
+                    }
+                }
+            }
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt]: I generally found this code quite hard to follow, not sure if there is a better way to do it. The comments help, but I kept getting confused when trying to double check what it was doing.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                // Check that no input states have been duplicated to fool our counting
+                encumbranceGroup.value.groupBy { it.first }.filter { it.value.size > 1 }.apply{
+                    this.forEach{ failureReasons.add(
+                        ContractVerificationFailureImpl(
+                            contractClassName = "",
+                            contractStateClassNames = emptyList(),
+                            exceptionClassName = IllegalArgumentException::class.java.canonicalName,
+                            exceptionMessage = "Encumbrance check failed: State ${statesByTx.key}, " +
+                                    "${it.key} is part " +
+                                    "is used ${it.value.size} times as input!"))
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt]: This looks a bit odd:
`... is part is ...`

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import net.corda.v5.crypto.SecureHash
+import net.corda.v5.ledger.utxo.ContractVerificationFailure
+import net.corda.v5.ledger.utxo.StateAndRef
+import java.lang.IllegalArgumentException
+
+
+private data class EncumbranceInfo(val stateIndex: Int, val encumbranceGroupSize: Int)
+
+
+@Suppress("NestedBlockDepth")
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionEncumbranceVerifier.kt]: Is this still needed?

**Review by vlajos — DISMISSED** [review]
It looks good to me.

**Review by lankydan — DISMISSED** [review]

**Review by vlajos — APPROVED** [review]

**Review by driessamyn — APPROVED** [review]
LGTM

**Review by mnesbit — COMMENTED** [review]
ok
